crates/runtime/src/search/types.rs

```diff
@@ -16,220 +16,66 @@ limitations under the License.
-            return Ok(vec![]);
-        }
-        let primary_keys_json = self.primary_keys_json()?;
-        let additional_columns_json = self.addition_columns_json()?;
-        let values = self.embedding_columns_list()?;
-        let scores = self.score_values()?;
-
-        values
-            .iter()
-            .enumerate()
-            .map(|(i, value)| {
-                let Some(score) = scores.get(i) else {
-                    return Err(Error::EmbeddingError {
-                        source: format!("No distance returned for {i}th result").into(),
-                    });
-                };
-
-                Ok(Match {
-                    value: value.clone(),
-                    score: *score,
-                    dataset: table.to_string(),
-                    primary_key: primary_keys_json.get(i).cloned().unwrap_or_default(),
-                    metadata: additional_columns_json.get(i).cloned().unwrap_or_default(),
-                })
-            })
-            .collect::<Result<Vec<Match>>>()
-    }
-}
-
-/// The results of [`CandidateGeneration::search`]'s on a single table.
-pub struct VectorSearchGenerationTableResult {
-    pub data: Vec<SendableRecordBatchStream>,
-    pub primary_keys: Vec<String>,
-}
-
-pub type VectorSearchResult = HashMap<TableReference, VectorSearchTableResult>;
+pub type VectorSearchResult = HashMap<TableReference, AggregationResult>;
 
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 #[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
 pub struct Match {
-    /// The value of the match (e.g., document snippet, identifier, etc.)
-    value: String,
+    /// The matches for this result
+    matches: HashMap<String, MatchType>,
+
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    data: HashMap<String, Value>,
+
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    primary_key: HashMap<String, Value>,
 
     /// The similarity of the match to the query
     score: f64,
 
     /// The name of the dataset where the match was found
     dataset: String,
 
-    /// Primary key(s) identifying the matched item in the dataset
-    #[serde(skip_serializing_if = "HashMap::is_empty")]
-    pub primary_key: HashMap<String, serde_json::Value>,
-
-    /// Additional metadata for the match, requested explicitly by the user.
     #[serde(skip_serializing_if = "HashMap::is_empty")]
     pub metadata: HashMap<String, serde_json::Value>,
 }
 
-impl Match {
-    #[must_use]
-    pub fn value(&self) -> &str {
-        &self.value
+#[derive(Clone, PartialEq, Debug, Deserialize, Serialize)]
+#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
+#[serde(untagged)]
+pub enum MatchType {
+    Single(Value),
+    Multiple(Vec<Value>),
+}
+
+impl From<Vec<Value>> for MatchType {
+    fn from(mut value: Vec<Value>) -> Self {
+        if value.len() == 1 {
+            if let Some(v) = value.pop() {
+                return MatchType::Single(v);
+            }
```

Comment:
```suggestion
           let Some(v) = value.pop() else {
               unreachable!("The value array must have one element");
           };
           return MatchType::Single(v);
```

nit: this can be an infallible match because we checked the length